### PR TITLE
Assume Content-Type is json for endpoints that require json

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

#4568

## Proposed Changes

Create a custom filter that parses json in the request body by default without checking the content-type header. Apply the filter to endpoints that expect JSON in the request body.

## Additional Info

The built in warp json filter `warp::body::json` defaults to `Content-Type application/json` if a Content-Type isn't provided. If a `Content-Type` is provided and it is not `application/json` the request will fail.

I ran several tests on my own, and was unable to recreate a failed endpoint request when `Content-Type` was omitted. I was however able to recreate a failed endpoint request when `Content-Type` was included but wasn't `application/json`.

-----

In out test suite note that `builder.json()` appends `Content-Type application/json` to the header by default.

I created a new function `post_generic_json_without_content_type_header` which creates a post request with serialized json in the body without including `Content-Type application/json` in the header.

Locally I replaced all post requests in our test suite to NOT include the `Content-Type` header and all the tests passed. In this PR I only updated a single test to send a POST request without the `Content-Type` header.

I plan on also spinning up the client locally and making some curl requests myself to make sure I haven't missed anything.

### CODE REVIEW

**Comment by eserilev** [comment]
~~based on @michaelsproul work in #4595 it seems the GET validator/block endpoint fails when a content-type isn't provided~~

~~the weird part here is that our test suite doesn't provide a content type header when making this request~~. ~~furthermore, based on the warp filters for this request I dont see a place where we even require a content-type header~~. ~~Will investigate this further~~

EDIT: nevermind!

**Comment by michaelsproul** [comment]
@eserilev I don't think any of our GETs fail due to lack of `Content-Type`. In my PR I was referring to `POST /eth/v2/beacon/blocks`

**Comment by eserilev** [comment]
@michaelsproul I did some digging in the curl manual, apparently curl defaults to `Content-Type application/x-www-form-urlencoded` if no content-type header is provided.

> Like [-d, --data](https://curl.se/docs/manpage.html#-d) the default content-type sent to the server is application/x-www-form-urlencoded

https://curl.se/docs/manpage.html

 As i mentioned above, the default `warp:body:json` filter will always default to `Content-Type application/json` **unless** a different content type is provided. If a different content type is provided the request will fail. 

I believe my changes here should effectively ignore the content-type header and resolve issues users have when making curl requests  to relevant endpoints.

**Comment by eserilev** [comment]
By the way, per the HTTP spec, I think ignoring/overriding the content-type header in this situation is ok

>In practice, resource owners do not always properly configure their origin server to provide the correct Content-Type for a given representation. Some user agents examine the content and, in certain cases, override the received type (for example, see [[Sniffing](https://www.rfc-editor.org/rfc/rfc9110.html#Sniffing)])

https://www.rfc-editor.org/rfc/rfc9110.html#name-content-type

**Comment by chong-he** [comment]
I have tested this and for queries without `-H 'Content-Type: application/json'`, the queries work as intended.

Details: 

With `Lighthouse v4.3.0-9a24a01`, querying:
```bash
curl -X 'POST'   'http://localhost:5052/eth/v1/beacon/rewards/sync_committee/head'   -H 'accept: application/json'  -d '["502264"]' | jq
```
returns:
```json
{
  "execution_optimistic": true,
  "finalized": false,
  "data": []
}
```

With `Lighthouse v4.5.0-441fc16`, the same query returns:
```json
{
  "code": 400,
  "message": "BAD_REQUEST: Unsupported endpoint version: v1",
  "stacktraces": []
}
```

If this is good we can merge this PR?

File: common/eth2/src/lib.rs [hunk 1]
```diff
@@ -365,10 +365,30 @@ impl BeaconNodeHttpClient {
...
+        url: U,
+        body: &T,
+        timeout: Option<Duration>,
+    ) -> Result<Response, Error> {
+        let mut builder = self.client.post(url);
+        if let Some(timeout) = timeout {
+            builder = builder.timeout(timeout);
+        }
+
+        let serialized_body = serde_json::to_vec(body).unwrap();
```
Comment (michaelsproul) [inline common/eth2/src/lib.rs]:
```suggestion
        let serialized_body = serde_json::to_vec(body)?;
```

We can't unwrap here. We may need to `map_err` into some existing variant, or add a new variant.

File: common/warp_utils/src/json.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+use crate::reject;
+
+struct Json;
+
+type BoxError = Box<dyn StdError + Send + Sync>;
+
+impl Json {
+    fn decode<B: Buf, T: DeserializeOwned>(mut buf: B) -> Result<T, BoxError> {
+        serde_json::from_slice(&buf.copy_to_bytes(buf.remaining())).map_err(Into::into)
```
Comment (michaelsproul) [inline common/warp_utils/src/json.rs]: I think `Bytes` dereferences to `&[u8]`

```suggestion
        serde_json::from_slice(&buf).map_err(Into::into)
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looking good, would be nice to merge this for 4.6.0-rc

Looks like there's a cargo audit failure that could probably be fixed by `cargo update -p zerocopy`.

File: common/warp_utils/src/json.rs (hunk 2 above)
Comment (eserilev) [reply common/warp_utils/src/json.rs]: This is a `Buf` which I dont think deferences directly to `&[u8]`. I think the other option here is `&buf.chunk()`

Comment (eserilev) [reply common/warp_utils/src/json.rs]: ah I think I can just pass in bytes directly, I see what you're saying now

**Review by michaelsproul — APPROVED** [review]

**Comment by chong-he** [comment]
Did a testing on passing an invalid data. 

With Lighthouse `v4.5.0-`3e8f176`, querying:
(note the validator passed is an alphabet to provide invalid data)
```bash
curl -X 'POST'   'http://localhost:5052/eth/v1/beacon/rewards/sync_committee/head'   -H 'accept: application/json'  -d '[“a"]' | jq
```

returns:
```json
{
  "code": 400,
  "message": "BAD_REQUEST: Error(\"a cannot be parsed as a slot: invalid digit found in string\", line: 3, column: 1)",
  "stacktraces": []
}
```

With `Lighthouse v4.5.0-441fc16`, the same query returns:
```json
{
  "code": 400,
  "message": "BAD_REQUEST: Unsupported endpoint version: v1",
  "stacktraces": []
}
```

The updated code is able to show the correct error message when an invalid data is passed, which is good.

File: common/warp_utils/src/json.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    fn decode<T: DeserializeOwned>(bytes: Bytes) -> Result<T, BoxError> {
+        serde_json::from_slice(&bytes).map_err(Into::into)
+    }
+}
+
+pub fn json<T: DeserializeOwned + Send>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy {
+    warp::body::bytes().and_then(|bytes: Bytes| async move {
+        Json::decode(bytes).map_err(|err| reject::custom_bad_request(format!("{:?}", err)))
+    })
+}
```
Comment (dapplion) [inline common/warp_utils/src/json.rs:22]: Diff from `warp::filters::body::json`

```diff
pub fn json<T: DeserializeOwned + Send>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy {
-    is_content_type::<Json>()
        .and(bytes())
        .and_then(|buf| async move {
-           Json::decode(buf).map_err(|err| {
-               tracing::debug!("request json body error: {}", err);
-               reject::known(BodyDeserializeError { cause: err })
-           })
+           Json::decode(bytes).map_err(|err| reject::custom_bad_request(format!("{:?}", err)))
        })
}
```

```diff
impl Decode for Json {
-    fn decode<B: Buf, T: DeserializeOwned>(mut buf: B) -> Result<T, BoxError> {
+    fn decode<T: DeserializeOwned>(bytes: Bytes) -> Result<T, BoxError> {
-        serde_json::from_slice(&buf.copy_to_bytes(buf.remaining())).map_err(Into::into)
+        serde_json::from_slice(&bytes).map_err(Into::into)
    }
}
```

Comment (dapplion) [reply common/warp_utils/src/json.rs:22]: 👍  Equivalent code with the suggestion to remove the copy from https://github.com/sigp/lighthouse/pull/4575#discussion_r1430824300

**Review by dapplion — APPROVED** [review]
PR should be good to go once conflicts fixed

Does it slightly improve https://github.com/sigp/lighthouse/pull/5104 as there's one less filter?

File: common/warp_utils/src/json.rs (hunk 3 above)
Comment (dapplion) [reply common/warp_utils/src/json.rs:22]: @eserilev consider appending some extra message context like

https://github.com/sigp/lighthouse/blob/1be5253610dc8fee3bf4b7a8dc1d01254bc5b57d/common/warp_utils/src/reject.rs#L164-L165

File: common/warp_utils/src/reject.rs [hunk 4]
```diff
@@ -82,6 +82,15 @@ pub fn custom_bad_request(msg: String) -> warp::reject::Rejection {
...
 }
 
+#[derive(Debug)]
+pub struct CustomDeserializeError(pub String);
+
+impl Reject for CustomBadRequest {}
+
+pub fn custom_deserialize_error(msg: String) -> warp::reject::Rejection {
+    warp::reject::custom(CustomDeserializeError(msg))
+}
```
Comment (eserilev) [inline common/warp_utils/src/reject.rs]: added some extra message context

File: common/warp_utils/src/reject.rs [hunk 5]
```diff
@@ -161,6 +170,9 @@ pub async fn handle_rejection(err: warp::Rejection) -> Result<impl warp::Reply,
     if err.is_not_found() {
         code = StatusCode::NOT_FOUND;
         message = "NOT_FOUND".to_string();
+    } else if let Some(e) = err.find::<crate::reject::CustomDeserializeError>() {
+        message = format!("BAD_REQUEST: body deserialize error: {}", e);
+        code = StatusCode::BAD_REQUEST;
```
Comment (eserilev) [inline common/warp_utils/src/reject.rs]: added some extra message context

**Comment by eserilev** [comment]
conflicts resolved and added additional error message context

> Does it slightly improve https://github.com/sigp/lighthouse/pull/5104 as there's one less filter?

From what I understand #5104 is caused by nested futures. Since the json filter still makes an async call, I have a feeling it doesnt make any improvements. Tbh I dont know enough to give a definite answer though

**Review by dapplion — APPROVED** [review]
Thanks for the revisions!

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
@mergifyio queue

**Comment by michaelsproul** [comment]
@mergifyio requeue

**Pull request merged**